File: pathform2json.py
#!/usr/bin/env python3
"""
PathForm to JSON CLI Tool

Converts PathForm text files to JSON format.
"""

import sys
import json
import argparse
from pathform import parse_pathform, to_json


def main():
    parser = argparse.ArgumentParser(
        description='Convert PathForm text to JSON',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  pathform2json input.pf > output.json
  pathform2json input.pf -o output.json
  pathform2json input.pf --indent 4
  cat input.pf | pathform2json
        """
    )
    parser.add_argument(
        'input',
        nargs='?',
        type=argparse.FileType('r'),
        default=sys.stdin,
        help='Input PathForm file (default: stdin)'
    )
    parser.add_argument(
        '-o', '--output',
        type=argparse.FileType('w'),
        default=sys.stdout,
        help='Output JSON file (default: stdout)'
    )
    parser.add_argument(
        '--indent',
        type=int,
        default=2,
        help='JSON indentation (default: 2, use 0 for compact)'
    )
    parser.add_argument(
        '--compact',
        action='store_true',
        help='Output compact JSON (no indentation)'
    )

    args = parser.parse_args()

    try:
        # Read input
        text = args.input.read()
        
        # Parse and convert
        indent = 0 if args.compact else args.indent
        json_output = to_json(text, indent=indent)
        
        # Write output
        args.output.write(json_output)
        if args.output != sys.stdout:
            args.output.write('\n')
        
        return 0
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        return 1


if __name__ == '__main__':
    sys.exit(main())

File: pathform2json.js
#!/usr/bin/env node
/**
 * PathForm to JSON CLI Tool (Node.js)
 * 
 * Converts PathForm text files to JSON format.
 */

const fs = require('fs');
const path = require('path');
const { parsePathform, toJSON } = require('./pathform.js');

function main() {
    const args = process.argv.slice(2);
    
    // Simple argument parsing
    let inputFile = null;
    let outputFile = null;
    let indent = 2;
    let compact = false;
    
    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        if (arg === '-o' || arg === '--output') {
            outputFile = args[++i];
        } else if (arg === '--indent') {
            indent = parseInt(args[++i], 10);
        } else if (arg === '--compact') {
            compact = true;
            indent = 0;
        } else if (arg === '-h' || arg === '--help') {
            console.log(`
Usage: pathform2json [options] [input]

Convert PathForm text to JSON.

Options:
  -o, --output FILE    Output JSON file (default: stdout)
  --indent N           JSON indentation (default: 2, use 0 for compact)
  --compact             Output compact JSON (no indentation)
  -h, --help           Show this help message

Examples:
  pathform2json input.pf > output.json
  pathform2json input.pf -o output.json
  pathform2json input.pf --indent 4
  cat input.pf | pathform2json
            `);
            process.exit(0);
        } else if (!arg.startsWith('-')) {
            inputFile = arg;
        }
    }
    
    try {
        // Read input
        const text = inputFile 
            ? fs.readFileSync(inputFile, 'utf8')
            : fs.readFileSync(0, 'utf8');
        
        // Parse and convert
        const jsonOutput = toJSON(text, compact ? 0 : indent);
        
        // Write output
        if (outputFile) {
            fs.writeFileSync(outputFile, jsonOutput + '\n', 'utf8');
        } else {
            process.stdout.write(jsonOutput);
        }
        
        process.exit(0);
    } catch (error) {
        console.error(`Error: ${error.message}`);
        process.exit(1);
    }
}

if (require.main === module) {
    main();
}

